src/Sarif.Viewer.VisualStudio/Controls/InfoBar.cs

```diff
@@ -45,32 +46,47 @@ public class InfoBar : IVsInfoBarUIEvents
         private IVsInfoBarUIElement uiElement;
         private uint eventCookie;
 
+        private static readonly ReaderWriterLockSlimWrapper s_infoBarLock = new ReaderWriterLockSlimWrapper(new ReaderWriterLockSlim());
+        private static readonly IDictionary<InfoBar, ExceptionalConditions> s_infoBarToConditionDictionary = new Dictionary<InfoBar, ExceptionalConditions>();
```

Comment:
In a multi-thread environment, ConcurrentDictionary<> may be better to use.

Reply:
I didn't think of that. Thanks for suggesting it. I'm not sure it will work, though.

Notice the four calls to `AddInfoBarIfRequiredAsync`  below. I want the whole _set_ of operations to be atomic. Also, within each call to `AddInfoBarIfRequiredAsync`, I want the logic that checks whether the specified bit is set to be atomic with the dictionary update.

Does this make sense? Given that, do you think I can replace the lock with a `ConcurrentDictionary`?

---
In reply to: [527217732](https://github.com/microsoft/sarif-visualstudio-extension/pull/278#discussion_r527217732) [](ancestors = 527217732)

Reply:
Come to think of it, the sequence of four calls does _not_ have to be atomic. If I add one error bar, and _you_ add the one I was _about_ to add before I get around to it, then no harm is done. When I try to add it, I will find that it's already there.

Let me look more closely at the internals of `AddInfoBarIfRequiredAsync` to see if its entire body needs to be atomic, or whether I really can get away with using `ConcurrentDictionary`.

---
In reply to: [527233531](https://github.com/microsoft/sarif-visualstudio-extension/pull/278#discussion_r527233531) [](ancestors = 527233531,527217732)

Reply:
> Come to think of it, the sequence of four calls does _not_ have to be atomic. If I add one error bar, and _you_ add the one I was _about_ to add before I get around to it, then no harm is done. When I try to add it, I will find that it's already there.
> 
> Let me look more closely at the internals of `AddInfoBarIfRequiredAsync` to see if its entire body needs to be atomic, or whether I really can get away with using `ConcurrentDictionary`.
> 
> In reply to: [527233531](https://github.com/microsoft/sarif-visualstudio-extension/pull/278#discussion_r527233531) [](ancestors = 527233531,527217732)

IMO, this is not required, but a preferred data type to use. Thanks for taking this seriously. 👍

Reply:
I always take PR feedback seriously :-) So let me think out load for a moment.

Suppose it turns out that the dictionary accesses need to be encapsulated in a higher-level lock (for example, locking the entire body of `AddInfoBarIfRequiredAsync`, or locking an entire _sequence_ of calls to that method, which is what I'm currently doing. In that case, is there any benefit to using `ConcurrentDictionary`? It would seem that you incur the perf overhead of a concurrent collection without getting any additional safety back.

---
In reply to: [527238898](https://github.com/microsoft/sarif-visualstudio-extension/pull/278#discussion_r527238898) [](ancestors = 527238898)

Reply:
You do come with a good point about  two options here, i.e., using ConcurrentDictionary vs executing locks, and they can bring different performance metrics. And this is a VS extension, which cares more about user experience than some server services.

My original thought is since this is a multi-thread environment, why not using a safer data type to avoid issue caused by a  later change, which slips out of the lock.

So, I agree that the current implementation doesn't need ConcurrentDictionary<> type. For potential perf issues, we don't have to change.  :)

Looks this is a very useful discussion.

Reply:
Ok, you've convinced me. Perf is not an issue here, because we are

(a) performing a small number of operations (there are only a few info bars), and
(b) performing them on a human time scale (whenever a build finishes, for example).

So it comes down to safety, and you are right: On the day somebody performs a dictionary operation _outside_ of the lock -- wrong though that may be -- we will find ourselves with a hard to diagnose bug.

So I will make the change.

---
In reply to: [527253972](https://github.com/microsoft/sarif-visualstudio-extension/pull/278#discussion_r527253972) [](ancestors = 527253972)

Reply:
> Ok, you've convinced me. Perf is not an issue here, because we are
> 
> (a) performing a small number of operations (there are only a few info bars), and
> (b) performing them on a human time scale (whenever a build finishes, for example).
> 
> So it comes down to safety, and you are right: On the day somebody performs a dictionary operation _outside_ of the lock -- wrong though that may be -- we will find ourselves with a hard to diagnose bug.
> 
> So I will make the change.
> 
> In reply to: [527253972](https://github.com/microsoft/sarif-visualstudio-extension/pull/278#discussion_r527253972) [](ancestors = 527253972)

👍 Good talk! Thanks.

Reply:
Fixed.

---
In reply to: [527263211](https://github.com/microsoft/sarif-visualstudio-extension/pull/278#discussion_r527263211) [](ancestors = 527263211)